### ISSUE TITLE
Zip push deployment timeout

### ISSUE THREAD

**Issue by IlyaFinkelshteyn** [projectkudu/kudu#2700]
We recently implemented support for Zip Push Deploy as part of our AppVeyor CI service:  https://www.appveyor.com/docs/deployment/azure-app-service-zip-push-deploy/

It works great, but some customers are getting timeout error while actual deployment is successful: https://github.com/appveyor/ci/issues/2075#issuecomment-365511155

We have 5 minutes timeout hardcoded in `HttpClient` instance which is doing Zip Push Deploy `POST`. The same setting works great for years for WebJobs publishing with Kudu API, but apparently with Zip Push deployments it is too short (if not in Async mode). 

I wonder if you have any statistics how long Zip Push deployment can take in reality. We will adjust our timeout and maybe even expose this setting in UI/YAML to make it configurable in the range you will advise.

**Comment by xt0rted** [comment]
Some info on the failure I received, it was a 511 KB zip file which contained 11 folders and a total of 13 files. The deploy was for a node functions app and used [Azure/azure-functions-pack](https://github.com/Azure/azure-functions-pack) to reduce the size of the deployment. It was also deployed to a staging slot with auto swap enabled. If it'd help I can get the info from when this deploy happened so the logs can be reviewed.

**Comment by davidebbo** [comment]
@xt0rted that's tiny and should have been fast. Do you always get slow times with that file, or was it a random occurrence?

- If always, please share a repro zip
- If random, we may need the app name and UTC time to look at our logs (the more recent the better)

**Comment by xt0rted** [comment]
As far as I know it was only 1 time. It happened on `eei-slack-notifications` in the `staging` slot and was on 2/14 around 12:44 AM EST.

**Comment by davidebbo** [comment]
The reason it was so slow is that it had to delete a massive number of files first (about 21,000) before it could copy the new files. That is why it was so slow.

I think this happened as you were transitioning from non-funcpack to funcpack, and many node modules had to go. If you look at the deployment log, that is clear.

**Comment by IlyaFinkelshteyn** [comment]
@davidebbo should we make Async to be a default setting to avoid situations like this?

**Comment by xt0rted** [comment]
You're right. All of my deploys up until that point were to the staging slot which went fine. Then earlier in the day I did my first official deploy & swap. So when I did my next deploy the old production slot is where the new deploy went to and that still had the `node_modules` folder.

This feels more like an edge case than the norm but using the async model could work around this entirely as long as AppVeyor polls that endpoint to let the user know how the deploy went.

**Comment by IlyaFinkelshteyn** [comment]
@davidebbo I see, will still leave option to disable Async because sometime people assume that they can run E2E tests (or something else which assumes site is up with a new bits) right after deployment.

**Comment by davidebbo** [comment]
@IlyaFinkelshteyn when using the async pattern, you get a URL that you can probe for progress. So using async does not imply fire-and-forget. You very much should still wait until it is complete. But you do this is a way that won't hit the timeout.

**Comment by IlyaFinkelshteyn** [comment]
@davidebbo thanks a lot, will do that!
